### PR TITLE
multi-loss optimization by adding a DownpourOpt worker

### PR DESCRIPTION
**Pull request by yaoxuefeng6**
This pr add a designed new version of downpouropt worker to get a batter speed performance when there are multi losses in a program.
in Python front-end， we fix _minimize function of distributed optimizer to support multi-loss in one program. To get better speed performance when there are multi losses, we add some additional attribute of sparse table:
1, set a sparse table "is_local" to pull all feasigns embedding before a training pass from parameter server and generate a local table, while training, the embeddings can be directly pull from this local table. This attribute can be used when a sparse table is only used for ff.
2, set a sparse table "is_async" to asynchronously run independent ops while pulling this sparse table's embeddings. other independent ops can be run when pull sparse asynchronously in worker.
3, Use DownpourOpt worker to adjust ops order to gather forward ops and backwards ops by optimized loss in program. So that we can run ops flexibly in this worker. Currently, this worker has good support on these two attribute as mentioned above. 
example:
``` python
adam = fluid.optimizer.Adam(learning_rate=0.000005)
adam = fleet.distributed_optimizer(adam, strategy={"device_worker":"DownpourSGDOpt"})
```

### CODE REVIEW

File: cmake/inference_lib.cmake [hunk 1]
```diff
@@ -150,6 +150,36 @@ endfunction()
 set(inference_lib_deps third_party paddle_fluid paddle_fluid_c paddle_fluid_shared paddle_fluid_c_shared)
 add_custom_target(inference_lib_dist DEPENDS ${inference_lib_deps})
 
+set(dst_dir "${FLUID_INFERENCE_INSTALL_DIR}/third_party/eigen3")
```
Comment (guru4elephant) [inline cmake/inference_lib.cmake]: why do you add these dependencies?

File: paddle/fluid/framework/data_set.cc [hunk 2]
```diff
@@ -638,6 +644,87 @@ int DatasetImpl<T>::ReceiveFromClient(int msg_type, int client_id,
...
+  int channel_num = multi_output_channel_.size();
+  if (read_thread_num < channel_num) {
+    read_thread_num = channel_num;
+  }
+  std::vector<std::thread> threads(read_thread_num);
+  consume_task_pool_.resize(consume_thread_num);
+  for (size_t i = 0; i < consume_task_pool_.size(); i++) {
+    consume_task_pool_[i].reset(new ::ThreadPool(1));
+  }
+  // std::vector<std::vector<std::future<void>>> task_futures(shard_num);
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.cc]: remove this comment if not used.

File: paddle/fluid/framework/data_set.cc [hunk 3]
```diff
@@ -638,6 +644,87 @@ int DatasetImpl<T>::ReceiveFromClient(int msg_type, int client_id,
...
+  // auto consume_func = [&local_map_tables]
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.cc]: remove comments.

File: paddle/fluid/framework/data_set.cc [hunk 4]
```diff
@@ -638,6 +644,87 @@ int DatasetImpl<T>::ReceiveFromClient(int msg_type, int client_id,
...
+  auto consume_func = [&local_map_tables](int shard_id,
+                                          std::vector<uint64_t>& keys) {
+    for (auto k : keys) {
+      if (local_map_tables[shard_id].find(k) ==
+          local_map_tables[shard_id].end()) {
+        local_map_tables[shard_id][k] = std::vector<float>(11, 0);
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.cc]: magic number should be avoid.

File: paddle/fluid/framework/data_set.cc [hunk 5]
```diff
@@ -638,6 +644,87 @@ int DatasetImpl<T>::ReceiveFromClient(int msg_type, int client_id,
...
+    std::vector<Record> vec_data;
+    std::vector<std::vector<uint64_t>> task_keys(shard_num);
+    std::vector<std::future<void>> task_futures;
+    this->multi_output_channel_[i]->Close();
+    this->multi_output_channel_[i]->ReadAll(vec_data);
+    for (size_t j = 0; j < vec_data.size(); j++) {
+      for (auto& feature : vec_data[j].uint64_feasigns_) {
+        int shard = feature.sign().uint64_feasign_ % shard_num;
+        task_keys[shard].push_back(feature.sign().uint64_feasign_);
+        // this->local_tables_[shard].insert(feature.sign().uint64_feasign_);
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.cc]: remove this comment if not used.

File: paddle/fluid/framework/data_set.cc [hunk 6]
```diff
@@ -638,6 +644,87 @@ int DatasetImpl<T>::ReceiveFromClient(int msg_type, int client_id,
...
+    threads[i] = std::thread(gen_func, i);
+  }
+  for (std::thread& t : threads) {
+    t.join();
+  }
+  for (size_t i = 0; i < consume_task_pool_.size(); i++) {
+    consume_task_pool_[i].reset();
+  }
+  consume_task_pool_.clear();
+  fleet_ptr_->PullSparseToLocalV2(table_id, feadim);
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.cc]: can you propose some other names rather than V2, I think this function name is not maintainable

File: paddle/fluid/framework/data_set.h [hunk 7]
```diff
@@ -179,6 +187,11 @@ class DatasetImpl : public Dataset {
   virtual int64_t GetMemoryDataSize();
   virtual int64_t GetShuffleDataSize();
   virtual void MergeByInsId() {}
+  virtual void GenerateLocalTablesUnlock(int table_id, int feadim,
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_set.h:190]: I think we need to refactor Dataset in the next version. Current interface of base class is too messy.

File: paddle/fluid/framework/device_worker.h [hunk 8]
```diff
@@ -207,54 +207,104 @@ class DownpourWorker : public HogwildWorker {
...
 
   std::shared_ptr<PullDenseWorker> _pull_dense_worker;
-  std::vector<::std::future<int32_t>> push_sparse_status_;
-  std::vector<::std::future<int32_t>> push_dense_status_;
+  // std::vector<::std::future<int32_t>> push_sparse_status_;
+  // std::vector<::std::future<int32_t>> push_dense_status_;
 
   // adjust ins weight
-  AdjustInsWeightConfig adjust_ins_weight_config_;
+  // AdjustInsWeightConfig adjust_ins_weight_config_;
```
Comment (guru4elephant) [inline paddle/fluid/framework/device_worker.h]: remove these comments if not used.

File: paddle/fluid/framework/device_worker.h [hunk 9]
```diff
@@ -207,54 +207,104 @@ class DownpourWorker : public HogwildWorker {
   void CopySparseTable();
   void CopyDenseTable();
   void CopyDenseVars();
-
- private:
-  bool need_dump_param_;
-  std::vector<std::string> dump_param_;
+  std::string PrintLodTensor(LoDTensor* tensor, int64_t start, int64_t end);
+  std::pair<int64_t, int64_t> GetTensorBound(LoDTensor* tensor, int index);
+  bool CheckValidOutput(LoDTensor* tensor, size_t batch_size);
```
Comment (guru4elephant) [inline paddle/fluid/framework/device_worker.h:212]: I think some variables can be declared in DownpourWorkerOpt.
btw, does the name of DownpourWorkerOpt has alternatives?

File: paddle/fluid/framework/trainer_desc.proto [hunk 10]
```diff
@@ -164,4 +165,7 @@ message TableParameter {
   optional int32 emb_dim = 10;
   optional int32 fea_dim = 11;
   optional string label_var_name = 12;
+  optional bool is_local = 13 [ default = false ];
+  optional bool is_async = 14 [ default = false ];
+  optional string async_wait_op_name = 15;
```
Comment (guru4elephant) [inline paddle/fluid/framework/trainer_desc.proto:172]: could you add some commend to this variable?

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/optimizer_factory.py [hunk 11]
```diff
@@ -251,90 +259,119 @@ def _minimize(self,
...
+                        program_configs[program_id]["pull_dense"].extend(
+                            [dense_table_index])
+                        program_configs[program_id]["push_dense"].extend(
+                            [dense_table_index])
+                    else:
+                        program_configs[program_id][
+                            "pull_dense"] = [dense_table_index]
+                        program_configs[program_id][
+                            "push_dense"] = [dense_table_index]
+                    '''
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/parameter_server/pslib/optimizer_factory.py]: remove this if not used.

**Review by guru4elephant — COMMENTED** [review]
very good PR. please clean some unused comments and rethink of some function and class names.

File: cmake/inference_lib.cmake (hunk 1 above)
Comment (yaoxuefeng6) [reply cmake/inference_lib.cmake]: Because some conflicts of older commits, will change it

File: paddle/fluid/framework/data_set.cc (hunk 4 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: will fix it

File: paddle/fluid/framework/data_set.cc (hunk 6 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.cc]: will fix it

File: paddle/fluid/framework/data_set.h (hunk 7 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/data_set.h:190]: Agree, will pay attention to it

File: paddle/fluid/framework/device_worker.h (hunk 9 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/device_worker.h:212]: All these variables are moved from private to protected. It should be clear without comments. working on alternatives

File: paddle/fluid/framework/device_worker.h (hunk 8 above)
Comment (yaoxuefeng6) [reply paddle/fluid/framework/device_worker.h]: will fix

**Review by fuyinno4 — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]
